modules/c/nrt/source/Utils.c

```diff
@@ -468,126 +468,163 @@ NRTAPI(char) nrt_Utils_cornersTypeAsCoordRep(nrt_CornersType type)
     return cornerRep;
 }
 
-static void normalize_dms(int* pDegrees, int* pMinutes, int* pSeconds)
+static int is_valid_dms(int degrees, int minutes, double seconds)
 {
-    /* Ensure seconds and minutes are still within valid range. */
-    *pMinutes += (*pSeconds / 60);
-    *pSeconds %= 60;
-
-    *pDegrees += (*pMinutes / 60);
-    *pMinutes %= 60;
-
-    *pDegrees %= 360;
-
-    const int minutes_to_seconds = 60;
-    const int degrees_to_seconds = 60 * minutes_to_seconds;
-    int total_seconds = (*pDegrees * degrees_to_seconds) + (*pMinutes * minutes_to_seconds) + *pSeconds;
-
-    *pDegrees = total_seconds / degrees_to_seconds;
-    total_seconds -= (*pDegrees * degrees_to_seconds);
-
-    *pMinutes = total_seconds / minutes_to_seconds;
-    total_seconds -= (*pMinutes * minutes_to_seconds);
-
-    *pSeconds = total_seconds;
-}
+    // these are maximums; values for lat/lon may be smaller
+    if (abs(degrees) >= 360)
+    {
+        return 0;
+    }
+    if (abs(minutes) >= 60)
+    {
+        return 0;
+    }
+    if (fabs(seconds) >= 60.0)
+    {
+        return 0;
+    }
 
-static void adjust_dms(int* pDegrees, int* pMinutes, int* pSeconds, char* pDir,
-    char positive_dir, char negative_dir)
-{
-    *pDir = positive_dir;
-    if (*pDegrees < 0)
+    // A negative value is allowed for the first non-zero value
+    if (degrees != 0)
     {
-        *pDir = negative_dir;
-        *pDegrees *= -1;
+        if ((minutes < 0) || (seconds < 0.0))
+        {
+            return 0;
+        }
     }
-    if (*pMinutes < 0)
+    else if (minutes != 0)
     {
-        *pDir = negative_dir;
-        *pMinutes *= -1;
+        if (seconds < 0.0)
+        {
+            return 0;
+        }
     }
-    if (*pSeconds < 0)
+
+    return 1;
+}
+
+static int is_valid_lat_dms(int degrees, int minutes, double seconds)
+{
+    if (abs(degrees) > 90)
     {
-        *pDir = negative_dir;
-        *pSeconds *= -1;
+        return 0;
     }
+    return is_valid_dms(degrees, minutes, seconds);
 }
 
 NRTPROT(void) nrt_Utils_geographicLatToCharArray(int degrees, int minutes,
-                                                 double seconds_, char *buffer7)
+                                                 double seconds, char *buffer7)
 {
-    int seconds = (int) round(seconds_);
+    assert(buffer7 != NULL);
+    char dir = 'N';
 
-    normalize_dms(&degrees, &minutes, &seconds);
-    const int max_degrees = 90;
-    while ((degrees > max_degrees) || (degrees < -max_degrees))
+    if (is_valid_lat_dms(degrees, minutes, seconds))
     {
-        if (degrees > max_degrees)
+        if (degrees <= 0)
         {
-            degrees = max_degrees - (degrees - max_degrees); // 91 = 89
+            if (degrees < 0)
```

Comment:
Can you pull out the common logic from `nrt_Utils_geographicLatToCharArray` and `nrt_Utils_geographicLonToCharArray` into a common function like the below use that in both places?

```suggestion

static void fix_dir_dms(char newdir, char* dir, int *degrees, int *minutes, double *seconds)
{
        int d = *degrees;
        int m = *minutes;
        double s = *seconds
        char cd = *dir

        if (d < 0)
        {
            cd = newdir;
            d *= -1;
        }
        else if (m < 0)
        {
            cd = newdir;
            m *= -1;
        }
        else if (m == 0 && s < 0)
        {
            cd = newdir;
            s *= -1;
        }
        
        s = round(s);

        /* Ensure seconds and minutes are still within valid range. */
        if (s >= 60.0)
        {
            s -= 60.0;

            if (++m >= 60)
            {
                m -= 60;
                ++m;
            }
        }
        
        *dir = cd;
        *degrees = d;
        *minutes = m;
        *seconds = s;
}

// example
fix_dir_dms('S', &dir, &degrees, &minutes, &seconds);
            
```

Reply:
Sure, that could be done ... but it's a change that's mostly unrelated to this particular fix.

Reply:
It's probably not clear from the diffs/submissions, but I first rolled back to NITRO 2.9.0 and then fixed the specific problem.  The duplicated code is from 2.9.0, not my specific fix.